examples/vimco.py

```diff
@@ -0,0 +1,385 @@
+eq_samples = args.eq_samples    # number of samples for the expectation
+                                # over E_q(z|x)
+lr = args.lr
+anneal_lr_factor = args.anneal_lr_factor
+anneal_lr_epoch = args.anneal_lr_epoch
+batch_norm = args.batch_norm == 'true' or args.batch_norm == 'True'
+res_out = args.outfolder
+nonlin_enc = get_nonlin(args.nonlin_enc)
+nonlin_dec = get_nonlin(args.nonlin_dec)
+nhidden = args.nhidden
+latent_size = args.nlatent
+dataset = args.dataset
+batch_size = args.batch_size
+num_epochs = args.nepochs
+eval_epoch = args.eval_epoch
+
+assert dataset in ['sample','fixed'], "dataset must be sample|fixed"
+
+np.random.seed(1234) # reproducibility
+
+### SET UP LOGFILE AND OUTPUT FOLDER
+if not os.path.exists(res_out):
+    os.makedirs(res_out)
+
+# write commandline parameters to header of logfile
+args_dict = vars(args)
+sorted_args = sorted(args_dict.items(), key=operator.itemgetter(0))
+description = []
+description.append('######################################################')
+description.append('# --Commandline Params--')
+for name, val in sorted_args:
+    description.append("# " + name + ":\t" + str(val))
+description.append('######################################################')
+
+shutil.copy(os.path.realpath(__file__), os.path.join(res_out, filename_script))
+logfile = os.path.join(res_out, 'logfile.log')
+model_out = os.path.join(res_out, 'model')
+with open(logfile,'w') as f:
+    for l in description:
+        f.write(l + '\n')
+
+
+sym_iw_samples = T.iscalar('iw_samples')
+sym_eq_samples = T.iscalar('eq_samples')
+sym_lr = T.scalar('lr')
+sym_x = T.matrix('x')
+
+
+def bernoullisample(x):
+    return np.random.binomial(1,x,size=x.shape).astype(theano.config.floatX)
+
+
+### LOAD DATA AND SET UP SHARED VARIABLES
+if dataset is 'sample':
+    print "Using real valued MNIST dataset to binomial sample dataset after every epoch "
+    train_x, train_t, valid_x, valid_t, test_x, test_t = load_mnist_realval()
+    del train_t, valid_t, test_t
+    preprocesses_dataset = bernoullisample
+else:
+    print "Using fixed binarized MNIST data"
+    train_x, valid_x, test_x = load_mnist_binarized()
+    preprocesses_dataset = lambda dataset: dataset #just a dummy function
+
+train_x = np.concatenate([train_x,valid_x])
+
+train_x = train_x.astype(theano.config.floatX)
+test_x = test_x.astype(theano.config.floatX)
+
+num_features=train_x.shape[-1]
+
+sh_x_train = theano.shared(preprocesses_dataset(train_x), borrow=True)
+sh_x_test = theano.shared(preprocesses_dataset(test_x), borrow=True)
+
+
+### MODEL SETUP
+# Recognition model q(z|x)
+l_in = lasagne.layers.InputLayer((None, num_features))
+l_enc_h1 = lasagne.layers.batch_norm(DenseLayer(l_in, num_units=nhidden, name='Q_DENSE1'), name='Q_')
+l_enc_h1 = lasagne.layers.batch_norm(DenseLayer(l_enc_h1, num_units=nhidden, name='Q_DENSE2'),name='Q_')
+l_enc_h1 = lasagne.layers.batch_norm(DenseLayer(l_enc_h1, num_units=nhidden, name='Q_DENSE3'),name='Q_')
```

Comment:
Is it normal the hidden layers do not have nonlinearities?
(If so, `nonlin_enc`, `nonlin_dec` are not used)